torch/_inductor/ir.py

```diff
@@ -5261,3 +5262,97 @@ def codegen_collective(self, wrapper, output_name, input_names):
+        inputs_allow_none,
+    ):
+        super().__init__(layout, inputs, outputs, constant_args)
+        self.inputs_allow_none = inputs_allow_none
+
+    @classmethod
+    def create(
+        cls,
+        x: "TensorBox",
+        output_split_sizes: Optional["TensorBox"],
+        input_split_sizes: Optional["TensorBox"],
+        tag: str,
+        ranks: List[int],
+        group_size: int,
+    ):
+        x_realized = cls.realize_input(x)
+        inputs_allow_none = []
+        output_split_sizes_realized = None
+        if output_split_sizes is not None:
+            output_split_sizes_realized = cls.realize_input(output_split_sizes)
+            # NCCL/GLOO alltoall_base also assumes output_split_sizes to be int64,
+            # so hardcoding int64 here should be okay.
+            V.graph.name_to_coerced_dtype[output_split_sizes_realized.data.name] = torch.int64
+        input_split_sizes_realized = None
+        if input_split_sizes is not None:
+            input_split_sizes_realized = cls.realize_input(input_split_sizes)
+            # NCCL/GLOO alltoall_base also assumes input_split_sizes to be int64,
+            # so hardcoding int64 here should be okay.
+            V.graph.name_to_coerced_dtype[input_split_sizes_realized.data.name] = torch.int64
+        inputs_allow_none = [x_realized, output_split_sizes_realized, input_split_sizes_realized]
+        inputs = [inp for inp in inputs_allow_none if inp is not None]
+
+        outputs = []
+        new_size = x_realized.get_size()
+        output_split_sizes_sum_s = None
+        if output_split_sizes is not None:
+            output_split_sizes_sum_s = V.graph.current_node.meta['val'].size()[0]
+            new_size[0] = output_split_sizes_sum_s
+            # FlexibleLayout ctor used below needs to know which sizevar in output size
+            # is okay to skip the `isinstance(s, (Expr, int))` check for.
+            V.graph.data_dep_symbol_to_expr_str[str(output_split_sizes_sum_s)] = None
+
+        buff = OutputBuffer(
+            layout=FlexibleLayout(
+                device=x_realized.get_device(),
+                dtype=x_realized.get_dtype(),
+                size=new_size,
+            ),
+        )
+        outputs.append(buff)
+
+        layout = MultiOutputLayout(inputs[0].get_device())
+
+        packed = AllToAllSingle(
+            layout=layout,
+            inputs=inputs,
+            outputs=outputs,
+            constant_args=[tag, ranks, group_size],
+            inputs_allow_none=inputs_allow_none,
+        )
+
+        return cls.create_output_nodes(packed, outputs)[0]
+
+    def codegen_collective(self, wrapper, output_name, input_names):
+        # When output_split_sizes is specified, output shape of all_to_all_single is
+        # data-dependent on output_split_sizes, so we store the shape computation formula
+        # and during codegen time write this shape computation line before output buffer allocation.
```

Comment:
I'm not really a fan of organizing the code this way. Too much action at a distance. I think I still prefer for the meta function to perform the allocation, and then we automatically insert write outs for all unbacked ints the first time they come into scope. (This would also be a good place to add runtime asserts, which we need to add soon.)

Reply:
Removed the need for `data_dep_symbol_to_expr_str`